juniper_subscriptions/src/lib.rs

```diff
@@ -0,0 +1,446 @@
+//! This crate supplies [`SubscriptionCoordinator`] and
+//! [`SubscriptionConnection`] implementations for the
+//! [juniper](https://github.com/graphql-rust/juniper) crate.
+//!
+//! You need both this and `juniper` crate.
+//!
+//! [`SubscriptionCoordinator`]: juniper::SubscriptionCoordinator
+//! [`SubscriptionConnection`]: juniper::SubscriptionConnection
+
+#![deny(missing_docs)]
+#![deny(warnings)]
+#![doc(html_root_url = "https://docs.rs/juniper_subscriptions/0.14.2")]
+
+use std::{borrow::BorrowMut as _, iter::FromIterator, pin::Pin};
+
+use futures::{task::Poll, Stream};
+use juniper::{
+    http::{GraphQLRequest, GraphQLResponse},
+    BoxFuture, ExecutionError, GraphQLError, GraphQLSubscriptionType, GraphQLTypeAsync, Object,
+    ScalarValue, SubscriptionConnection, SubscriptionCoordinator, Value, ValuesStream,
+};
+
+/// Simple [`SubscriptionCoordinator`] implementation:
+/// - contains the schema
+/// - handles subscription start
+pub struct Coordinator<'a, QueryT, MutationT, SubscriptionT, CtxT, S>
+where
+    S: ScalarValue + Send + Sync + 'static,
+    QueryT: GraphQLTypeAsync<S, Context = CtxT> + Send + Sync,
+    QueryT::TypeInfo: Send + Sync,
+    MutationT: GraphQLTypeAsync<S, Context = CtxT> + Send + Sync,
+    MutationT::TypeInfo: Send + Sync,
+    SubscriptionT: GraphQLSubscriptionType<S, Context = CtxT> + Send + Sync,
+    SubscriptionT::TypeInfo: Send + Sync,
+    CtxT: Send + Sync,
+{
+    root_node: juniper::RootNode<'a, QueryT, MutationT, SubscriptionT, S>,
+}
+
+impl<'a, QueryT, MutationT, SubscriptionT, CtxT, S>
```

Comment:
This is a local `Coordinator` implementation. It requires `<QueryT, MutationT, SubscriptionT, CtxT, S>` to be able to hold `RootNode`.

Users can avoid implementing `SubscriptionCoordinator` if they'll depend on `juniper_subscriptions::Coordinator`

Reply:
Yes, but if we expect people to be implementing these things it might be better to create a proc macro in the future. Similar to how juniper's macros abstract out the juniper boilerplate even though you can still implement the boilerplate without the macros directly.